gitpod-io/gitpod
# Fix double submission on SSO Config modal

## Description

This work is intended to address a double submit issue on the SSO Config modal. Digging in, I think we can make some adjustments to how we use `Modal`s to more easily avoid this issue. In summary, these are the primary changes in this PR:

* Removed `Modal.onEnter`: No more listening to the `Enter` key manually and calling `onClose` based on the response of callback.
* Added `Modal.onSubmit`: utilizes a `<form>` to capture submission on `Enter` w/ native behavior. 
* Updated `ConfirmationModal` to account for the Modal change. It will also now put the confirm button into a `loading` state while `await`ing the `onConfirm` callback.

A callout here is if you pass an `onSubmit` to a Modal, you must have `<Button htmlType="submit">` in the `Modal` to have things like submit on Enter to work, which is typically your "Save" style button. This is how forms work in html though, so it feels correct.

I've also add a fixed width for the Modals above small screen sizes. A previous change to Modals seems to have allowed them to shrink a bit narrower than they used to be.

### A more in-depth explanation/opinion

The Modal `onEnter` prop is a bit problematic with how it listens to an `Enter` keypress and fires a handler/closes the modal. This leads to situations where `Enter` is pressed on a button, and it may or may not do what you want (i.e. a cancel vs. a save), and easy potential for double-fires if you’ve got an `onEnter` on the Modal, and an `onClick` on a Button. We work around this in a few places, but I think a better pattern is to rely on html `<form>` native behavior for submit on `Enter` functionality.

I also think we shouldn’t automatically fire `onClose`, and leave that up to the consumer of the Modal to do in their `onSubmit` handler. I've found that having a boolean return control that behavior can end up leaving a lot of room for bugs where one might not return the right value all the time. Consumers calling `onClose` is quite a bit more explicit.

## Related Issue(s)

Fixes WEB-397

## How to test

As this PR effects all of our Modals, I've spent quite a bit of time testing them to make sure they keep working as expected. There were a few opportunities to improve some of them a little, and fix a few bugs along the way.

Testing out any of the Modals in the product is the best way to verify behavior w/ these changes.

## Documentation


#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - brad-web-31eab1565e2</li>
	<li><b>🔗 URL</b> - <a href="https://brad-web-31eab1565e2.preview.gitpod-dev.com/workspaces" target="_blank">brad-web-31eab1565e2.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold
